### PR TITLE
AG 57 58 onbaording accessibility fixes

### PR DESCRIPTION
**Pull request by d4r1091**
[AG-57](https://ecosia.atlassian.net/browse/AG-57)
[AG-58](https://ecosia.atlassian.net/browse/AG-58)

## Context

Accessibility ticket created during [Engineering Offsite Core Journey Accessibility Review](https://ecosia.atlassian.net/wiki/spaces/DEV/pages/2627502081/Engineering+Offsite+April+18th-21st+2023+Spring+Cleaning#%E2%99%BF%EF%B8%8F%E2%9C%8A-Core-Journey-Accessibility-Review-(proposed-by-%40Zac-Colley-%26-%40sofiya.tepikin)).

Realized that the App's onboarding doesn't really provide a clear and concise explanation of the objects the steps are composed by.

## Approach

Analyzed the codebase I decided to take an approach that would fix all the onboarding sections, minimizing the codebase at UI level.
Grouped the onboarding section of each step in macro areas:
 - title
 - subtitle
 - content/illustration
 - cta button

Provided context via the `accessibilityLabel` primarily.
Whenever possible, added the `accessibilityValue` to voice over the intrinsic value of the object itself alongside a brief explanation via the `accessibilityLabel` itself.

## Other

Fixed an issue found in the Your Impact section, not counting the global number of trees planted.

### CODE REVIEW

File: Client/Ecosia/L10N/String.swift
Comment (lucaschifino) [inline Client/Ecosia/L10N/String.swift:1]: Really appreciate the thought of adding an image description 👏 I just wonder if therefore we would need an ok from someone from content 🤔 Also not sure if it won't be hard to maintain in the long run, for example I believe something might change on those images with the Google Market changes, which would then require new descriptions and therefore new translations. Have you given any thought to that?

TLDR: Appreciate the thought, but wonder if it might be better to leave the images out of the accessibility until we have an automatic or better defined process to come up with the descriptions

File: Client/Ecosia/L10N/String.swift [hunk 1]
```diff
@@ -182,7 +182,18 @@ extension String {
         case ecosiaLogoAccessibilityLabel = "Ecosia logo"
         case done = "Done"
         case findInPage = "Find in page"
+        case pageControlDots = "Page control dots"
+        case backButtonAccessibilityLabel = "Back button"
```
Comment (lucaschifino) [inline Client/Ecosia/L10N/String.swift:186]: I believe that an `UIButton` by default will be voiced as accessibility label + "button", so is the word button here needed? Even if it is not an `UIButton` you can also make it work as such by setting `.accessibilityTraits = .button`. I guess this might work best with how the user is used to accessibility in other apps.

File: Client/Ecosia/L10N/String.swift [hunk 2]
```diff
@@ -182,7 +182,18 @@ extension String {
...
+        case onboardingIllustrationTour1 = "This onboarding illustration shows how by performing searches via the Ecosia app, you are leveling up your tree planting impact score. A small search screenshot and a smaller Your Impact section is shown. Leaves on the background."
+        case onboardingIllustrationTour2 = "This onboarding illustration shows briefly an example of a before and after comparision of trees planted in a land. The image is a screenshot from the satellite view."
+        case onboardingIllustrationTour3 = "This onboarding illustration show a few numbers like the projects Ecosia is involved in, the total number of trees planted by the Ecosia community, alongisde the number of countries Ecosia is active. A small map of the planisphere with trees pins in few geographic location, background."
+        case onboardingIllustrationTour4 = "This onboarding illustration is a photo of a monkey climbing a tree. It function mainly as general decoration image."
+        case onboardingSkipTourButtonAccessibilityLabel = "Button that allows you to skip the onboarding"
```
Comment (lucaschifino) [inline Client/Ecosia/L10N/String.swift:193]: Same thing with regards the button's default behaviour here. Might not be the case but, for example, if the content of the button is "skip", it will already automatically voice over "skip button", which for me is a good label already. If not clear, you could then add an accessibility hint describing what the action does.

Like I said, not sure if it applies here, but if so I think the logic could be revised for all buttons, adding a very short accessibility label (potentially even only the button text) and an accessibility hint when not clear. I feel that is closer to what the user would be used in other apps.

**Review by lucaschifino — COMMENTED** [review]
Nice going with adding a lot of accessibility 👏 

Added some thoughts and comments, specially on default voice over behaviour, for you to consider, let me know what you think

File: Client/Ecosia/UI/NTP/Cells/NTPImpactCell.swift [hunk 3]
```diff
@@ -75,7 +75,7 @@ final class NTPImpactCell: UICollectionViewCell, NotificationThemeable, Reusable
         container.isAccessibilityElement = true
         container.shouldGroupAccessibilityChildren = true
         container.accessibilityHint = .localized(.yourImpactSectionAccessibilityHint)
-        container.accessibilityLabel = String(format: .localized(.yourImpactSectionAccessibilityLabel), "\(Int(currentProgress.value))", treesCount.text!)
+        container.accessibilityLabel = String(format: .localized(.yourImpactSectionAccessibilityLabel), globalCount.text!, treesCount.text!)
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/NTP/Cells/NTPImpactCell.swift:78]: That's right! It's indeed the other way around! Checked both counters and it spells indeed the values for both 👍 .
Will update the position. I guess they got misplaced whilst trying against different values.

File: Client/Ecosia/UI/Onboarding/WelcomeTourPlanet.swift [hunk 4]
```diff
@@ -23,6 +24,7 @@ final class WelcomeTourPlanet: UIView, NotificationThemeable {
 
         let topImage = UIImageView(image: .init(named: "tourSearch"))
         topImage.translatesAutoresizingMaskIntoConstraints = false
+        topImage.isAccessibilityElement = false
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/Onboarding/WelcomeTourPlanet.swift:27]: Seems like it's indeed the case! Must have been a refuse of the older attempts. Thanks for checking!

File: Client/Ecosia/UI/Onboarding/WelcomeTour.swift [hunk 5]
```diff
@@ -142,27 +146,32 @@ final class WelcomeTour: UIViewController,  NotificationThemeable {
...
+        labelStack.addArrangedSubview(subtitleLabel)
+        self.subtitleLabel = subtitleLabel
 
         let ctaButton = UIButton(type: .system)
         ctaButton.setTitle(.localized(.continueMessage), for: .normal)
         ctaButton.titleLabel?.font = .preferredFont(forTextStyle: .callout)
         ctaButton.titleLabel?.adjustsFontForContentSizeCategory = true
         ctaButton.translatesAutoresizingMaskIntoConstraints = false
         ctaButton.addTarget(self, action: #selector(forward), for: .primaryActionTriggered)
+        ctaButton.accessibilityLabel = .localized( isLastStep() ? .onboardingFinishCTAButtonAccessibilityLabel : .onboardingContinueCTAButtonAccessibilityLabel)
```
[the remark this answers is not in the document]
Comment (d4r1091) [reply Client/Ecosia/UI/Onboarding/WelcomeTour.swift:174]: I guess it can go yeah! Will remove the space in the next commit

File: Client/Ecosia/L10N/String.swift
Comment (d4r1091) [reply Client/Ecosia/L10N/String.swift:1]: Thanks for being on the same page with stressing at much these accessibility inputs.
I did indeed think twice or three times before adding it.
I thought that even if would be harder now to maintain due to the upcoming Google market changes that will affect inevitably our overall UX as well as the onboarding steps, the ultimate goal is to provide the best possible experience to the end user. Of course, that requires effort and most of the time we'll have to go after tradeoffs, etc...
At the same time, this has been a great opportunity to set the foundations of accessibility details that might make a difference.
Thought about a review from content, and that can be 💯 done. However, a lean approach to my perspective is to put something out there, then change it later. Obviously, I'm taking such initiative with accessibility, especially as it seems like a field where we can explore mostly by ourselves with an outcome that, even if not perfect, it's still a step towards the maximum we can do.
Reason for putting it now, ultimately, was to not forget the importance of it. With something coded in place, it will be easier to think of adding (and even removing it!) the accessible description/component. Would be much easily forgotten if otherwise.

File: Client/Ecosia/L10N/String.swift (hunk 2 above)
Comment (d4r1091) [reply Client/Ecosia/L10N/String.swift:193]: As some of the buttons have no text (like the back button), I thought about adding complementary info. Will remove it and make it `Skip the onboarding` perhaps.

File: Client/Ecosia/L10N/String.swift
Comment (ecotopian) [reply Client/Ecosia/L10N/String.swift:1]: Both sides present very valid arguments. I am leaning towards Dario's approach. Let's start simple and iterate on it. Also following our mindset to rather ask for forgiveness (if needed eventually) than for allowance. Those descriptions are not prominently user facing, so I am not concerned that there's misalignment. I find Dario's elaborations very fitting. We need to keep in mind, that we might need to change them though along imagery changes.

**Review by ecotopian — APPROVED** [review]
Looks way more accessible now :accessibility: 🙌
